### PR TITLE
adds tests to demonstrate various caveat use-cases

### PR DESCRIPTION
**Pull request by vroldanbet**
Implements various use-cases described in https://github.com/authzed/spicedb/issues/386

# What

implements various caveat use-cases to demonstrate viability of the design

# How

- IP Allowlists
- application attributes ([source](https://github.com/authzed/spicedb/issues/386#issuecomment-1163747750))
- subject created after resource ([source](https://github.com/authzed/spicedb/issues/386#issuecomment-1034278567))
- time-bound permission
- legal-guardian ([source](https://github.com/authzed/spicedb/issues/386#issuecomment-1252776047))

### CODE REVIEW

File: internal/dispatch/graph/computecheck_test.go [hunk 1]
```diff
@@ -369,6 +371,88 @@ func TestComputeCheckWithCaveats(t *testing.T) {
...
+
+					definition organization {
+						relation members: user
+						relation ip_allowlist_policy: organization#members | user:*
+					
+						permission policy = ip_allowlist_policy
+					}
+					
+					definition repository {
+						relation owner: organization | user
```
Comment (josephschorr) [inline internal/dispatch/graph/computecheck_test.go:386]: Drop the `| user` from the owner

File: internal/dispatch/graph/computecheck_test.go [hunk 2]
```diff
@@ -369,6 +371,88 @@ func TestComputeCheckWithCaveats(t *testing.T) {
 				},
 			},
 		},
+		{
+			"IP Allowlists example",
+			`definition user {}
+
+					definition organization {
+						relation members: user
+						relation ip_allowlist_policy: organization#members | user:*
```
Comment (josephschorr) [inline internal/dispatch/graph/computecheck_test.go:380]: Drop the `| user:*` from the relation

File: .github/workflows/lint.yaml [hunk 3]
```diff
@@ -49,7 +49,6 @@ jobs:
           working_directory: "tools/analyzers"
       - name: "Run custom analyzers"
         run: './tools/analyzers/analyzers -skip-pkg "github.com/authzed/spicedb/pkg/proto/dispatch/v1" -disallowed-nil-return-type-paths "*github.com/authzed/spicedb/pkg/proto/dispatch/v1.DispatchCheckResponse,*github.com/authzed/spicedb/pkg/proto/dispatch/v1.DispatchExpandResponse,*github.com/authzed/spicedb/pkg/proto/dispatch/v1.DispatchLookupResponse" ./...'
-      - uses: "authzed/actions/govulncheck@main"
```
Comment (josephschorr) [inline .github/workflows/lint.yaml:52]: Why is this removed?

File: internal/dispatch/graph/computecheck_test.go [hunk 4]
```diff
@@ -369,6 +371,373 @@ func TestComputeCheckWithCaveats(t *testing.T) {
...
+					"",
+				},
+			},
+		},
+		{
+			"authorize if resource was created before subject",
+			`definition root {
+						relation actors: actor
+					}
+                    definition resource {
```
Comment (josephschorr) [inline internal/dispatch/graph/computecheck_test.go:552]: mind fixing the formatting on these schema blocks?

File: internal/dispatch/graph/computecheck_test.go [hunk 5]
```diff
@@ -369,6 +371,373 @@ func TestComputeCheckWithCaveats(t *testing.T) {
...
+						relation reader: user
+						permission view = reader
+					}
+					
+					definition user {}`,
+			map[string]caveatDefinition{
+				"expired": {
+					"now() < timestamp(expiration)",
+					map[string]caveats.VariableType{
+						"expiration": caveats.StringType,
```
Comment (josephschorr) [inline internal/dispatch/graph/computecheck_test.go:616]: we have a TimestampType for dates and times: https://github.com/authzed/spicedb/blob/main/pkg/caveats/types.go#L29

Maybe add another test using it too?

File: internal/dispatch/graph/computecheck_test.go [hunk 6]
```diff
@@ -369,6 +371,373 @@ func TestComputeCheckWithCaveats(t *testing.T) {
...
+					  
+					  permission view = claimer + dependent_of
+					}
+					
+					definition user {
+						  relation dependent_of: user
+					}`,
+			map[string]caveatDefinition{
+				"legal_guardian": {
+					"age < 12 || (class != \"sensitive\" && age > 12 && age < 18)",
```
Comment (josephschorr) [inline internal/dispatch/graph/computecheck_test.go:669]: Use ` for the expression to remove the escaping

File: internal/dispatch/graph/computecheck_test.go [hunk 7]
```diff
@@ -369,6 +371,373 @@ func TestComputeCheckWithCaveats(t *testing.T) {
...
+						"class": "non-sensitive",
+					},
+				},
+				{
+					core.RelationTupleUpdate_CREATE,
+					"claim:sensitive_matter#dependent_of@user:daughter#dependent_of",
+					"legal_guardian",
+					map[string]any{
+						"age":   14,
+						"class": "sensitive",
```
Comment (josephschorr) [inline internal/dispatch/graph/computecheck_test.go:713]: I wonder if we should add an enum type of some kind

File: .github/workflows/lint.yaml (hunk 3 above)
Comment (vroldanbet) [reply .github/workflows/lint.yaml:52]: govulncheck was broken in `latest`. They don't have any releases, so I removed it. It's fixed now so I can revert this

File: internal/dispatch/graph/computecheck_test.go (hunk 7 above)
Comment (vroldanbet) [reply internal/dispatch/graph/computecheck_test.go:713]: Maybe be reasonable when we get to see the UX with schema parser in place

File: internal/dispatch/graph/computecheck_test.go (hunk 5 above)
Comment (vroldanbet) [reply internal/dispatch/graph/computecheck_test.go:616]: I added a test [here](https://github.com/authzed/spicedb/pull/878/files#diff-e576918c5584100ec96a2edd9fcf0b3048441257187a53378bd0ed720b910c1eR50-R67). I can change this test to use it, I just initially didn't know how to instantiate it but eventually figured it out

Comment (vroldanbet) [reply internal/dispatch/graph/computecheck_test.go:616]: We will address this in a follow up

**Review by josephschorr — APPROVED** [review]
LGTM
